File: countingSort.js
function countingSort(A){
	// Assume the max value 0th element, then we can search starting from 1st
	var k = A[0];
	var C = [];
	var B = [];
	// Finding maximum value in input array starting from 1st value
	for(var i=1; i<A.length; i++) if(A[i] > k) k = A[i];
	// Creating new array to use as big as the max value k and initializing to 0
	for(i=0; i<=k; i++) C[i] = 0;
	// Incrementing all the values that exists in input array
	for(var j=0; j<A.length; j++) C[A[j]]++;
	// Adding i-1 elements to i values
	for(i=1; i<=k; i++) C[i] += C[i-1];
	// Creating the output array B using the input array A and array C
	for(j=0; j<A.length; j++){
		B[C[A[j]]-1] = A[j];
		C[A[j]]--;
	}
	// Returning the output array B, we could take this as parameter and sort it in place.
	return B;
}

var arr = [3, 1, 0, 2, 5, 8, 6, 10, 9, 7, 4, 2];
var sortedArr = countingSort(arr);
console.log(sortedArr);

File: CountingSort.java
class CountingSort {
	
	public static int[] sort(int[] A){
		int[] B = new int[A.length];
		int k = A[0];
		for(int i=1; i<A.length; i++) if(k < A[i]) k = A[i];
		int[] C = new int[k+1];
		for(int j=0; j<=k; j++) C[j] = 0;
		for(int i=0; i<A.length; i++) C[A[i]]++;
		for(int j=1; j<=k; j++) C[j] += C[j-1];
		for(int i=0; i<A.length; i++){
			B[C[A[i]]-1] = A[i];
			C[A[i]]--;
		}
		return C;
	}
	
	public static void main(String[] args) {
		int[] arr = {3, 5, 1, 2, 9, 4, 8, 6, 7};
		for(int i=0; i<arr.length; i++) System.out.print(arr[i] + ", ");
		arr = sort(arr);
		System.out.println();
		for(int i=0; i<arr.length; i++) System.out.print(arr[i] + ", ");
	}
}